### PR TITLE
Don't create BvaDispatchTask if QualityReviewTask is open

### LINKED ISSUE #13416: `BvaDispatchTask` created before `QualityReviewTask` closes
## Description

Immediately after the JudgeDecisionReviewTask was completed at 2020-02-06 14:19:28, a QualityReviewTask was (correctly) created and eventually assigned on 2020-02-10.  Upon completion of the QR on 2020-02-11 21:31:59, a BvaDispatchTask was (correctly) created.
The problem is that a separate BvaDispatchTask was created earlier on 2020-02-06 20:19:36 for some unknown reason; BVALRENE is attributed with completing this BvaDispatchTask on 2020-02-06 20:46:21, which you can see in the Case Timeline that Dustin sent.

Root cause is still unknown after investigation.

## AC
- [ ] job that checks for `BvaDispatchTask` is created before `QualityReviewTask` closes

## Background/context/resources

From Dustin:
> The Board just reached out to me an appeal was assigned a dispatch task and dispatched despite there being open QR review tasks.  There is a concern the QR selection process might have a bug.  Should I just send this to batteam or is their a particular team or engineer with experience with the QR functionality that I should send it to?

https://dsva.slack.com/archives/C7P7M26QM/p1581540877063900?thread_ts=1581514491.056400&cid=C7P7M26QM

From Yoom:
>  I've spent many hours investigating this and also paired with Kevin and Kat looking for possible root causes. We couldn't find any other relevant appeal where a BvaDispatchTask was opened before a QualityReviewTask was closed, so it's not a recurring problem. I'll write up a ticket to create a periodic job that checks for such scenarios so that we can catch these problems hopefully before appeal decisions get dispatched.

## Technical notes

Find appeal where a `BvaDispatchTask` is created before a `QualityReviewTask` closes
```
qrAppIds=QualityReviewTask.all.map(&:appeal).uniq.pluck(:id)
dAppIds=BvaDispatchTask.where(appeal_id: qrAppIds).map(&:appeal).uniq.pluck(:id)
oddAppIds=Appeal.find(dAppIds).select{ |a| 
	qr_ids = a.tasks.where(type: :QualityReviewTask, parent_id: a.root_task.id).pluck(:id)
	d_ids = a.tasks.where(type: :BvaDispatchTask, parent_id: a.root_task.id).pluck(:id)
  id_pairs = qr_ids.product(d_ids)
  id_pairs.select{ |qr_id, d_id|
    Task.find(qr_id).closed_at - Task.find(d_id).created_at > 2
  }.present?
}.pluck(:id)
```

This results in appeals 19 and 12063. Appeal 12063 is the one Dustin brought up.

Appeal 19 is old and the BvaDispatchTask was created correctly after the child QualityReviewTask closed. For some reason, the parent QualityReviewTask closed a little later.

### Some notes from root cause investigation
```
Appeal 12063 (direct_review)                     ID     STATUS    ASGN_BY     ASGN_TO       CREATED_AT              CLOSED_AT
└── RootTask                                     282887 completed             Bva           2019-07-19 14:42:02 UTC 2020-02-06 20:46:21 UTC
    ├── TrackVeteranTask                         282888 completed             Vso           2019-07-19 14:42:03 UTC 2020-02-06 20:46:21 UTC
    ├── DistributionTask                         282889 completed             Bva           2019-07-19 14:42:03 UTC 2020-01-27 21:26:55 UTC
    │   └── InformalHearingPresentationTask      282890 completed             Vso           2019-07-19 14:42:03 UTC 2019-11-08 16:44:52 UTC
    │       └── InformalHearingPresentationTask  508938 completed VACOWENDET  BVAAMONTAGUE  2019-11-07 14:05:59 UTC 2019-11-08 16:44:52 UTC
    ├── JudgeAssignTask                          645151 completed             BVALYASUI     2020-01-27 21:26:55 UTC 2020-01-28 22:08:07 UTC
    ├── JudgeDecisionReviewTask                  649121 completed             BVAGWASIK     2020-01-28 22:08:07 UTC 2020-02-06 14:19:28 UTC
    │   └── AttorneyTask                         649122 completed BVALYASUI   VACOPRINSS    2020-01-28 22:08:07 UTC 2020-02-06 13:54:36 UTC
    ├── QualityReviewTask                        670894 completed             QualityReview 2020-02-06 14:19:29 UTC 2020-02-11 21:31:59 UTC
    │   └── QualityReviewTask                    676752 completed VACOPATRICN VACOPATRICN   2020-02-10 13:26:26 UTC 2020-02-11 21:31:59 UTC
    ├── BvaDispatchTask                          673035 completed             BvaDispatch   2020-02-06 20:19:36 UTC 2020-02-06 20:46:21 UTC
    │   └── BvaDispatchTask                      673036 completed             BVALRENE      2020-02-06 20:19:36 UTC 2020-02-06 20:46:21 UTC
    └── BvaDispatchTask                          683033 on_hold               BvaDispatch   2020-02-11 21:31:59 UTC
        ├── BvaDispatchTask                      683034 cancelled             BVADOSBORNE   2020-02-11 21:31:59 UTC 2020-02-12 15:28:08 UTC
        └── BvaDispatchTask                      684508 assigned  BVADOSBORNE BVAEJAMES     2020-02-12 15:28:08 UTC
```

For more, see [Echo thread](https://dsva.slack.com/archives/CJL810329/p1581541601271200)

Examining versions of each to find a matching timestamp for the errant task.

The task is "assigned_to_id: 19" by nil
`=> #<BvaDispatch id: 19, type: "BvaDispatch", name: "Board Dispatch", role: nil, url: "board-dispatch", participant_id: nil, created_at: nil, updated_at: nil, status: "active", status_updated_at: nil>`
The child task was marked completed by `whodunnit: nil` at 2020-02-06 20:46:21.446796997, 
about 26 minutes after it and its parent was created at 20:19:36.

It is curious to me that a different judge is reviewing the decision than the judge who assigned the case. The assigning judge BVALYASUI has no JudgeTeam.

### PR DESCRIPTION
**Pull request by yoomlam**
Resolves #13416
Prompted by Sanford's question: https://github.com/department-of-veterans-affairs/dsva-vacols/issues/178

### Description
Prevent creation of BvaDispatchTask if QualityReviewTask is open.

### Acceptance Criteria
- [ ] Code compiles correctly
- [x] `json` file has no PII

### Testing Plan
Mimic changed and new RSpecs in the Rails console. Identify any gaps in the logic or edge cases.
See https://github.com/department-of-veterans-affairs/dsva-vacols/issues/178#issuecomment-824073960

### CODE REVIEW

**Comment by yoomlam** [comment]
Related to PR #15040 that was not merged.

File: spec/feature/queue/judge_checkout_flow_spec.rb [hunk 1]
```diff
@@ -14,6 +17,86 @@
...
+      Organization.create!(id: 212, url: "bvajlmarch", name: "BVAJLMARCH")
+
+      # force skipping Quality Review (to mimic tab 2)
+      allow(QualityReviewCaseSelector).to receive(:select_case_for_quality_review?).and_return(false)
+    end
+    after { FeatureToggle.disable!(:special_issues_revamp) }
+    let!(:appeal) do
+      sji = SanitizedJsonImporter.from_file("spec/records/appeal-dispatch_before_quality_review_complete.json",
+                                            verbosity: 0)
+      sji.import
```
Comment (yoomlam) [inline spec/feature/queue/judge_checkout_flow_spec.rb:38]: Trying this out for the first time: https://github.com/department-of-veterans-affairs/caseflow/wiki/Exporting-and-Importing-Appeals
The JSON file is sanitized prod data.

File: app/models/tasks/bva_dispatch_task.rb [hunk 2]
```diff
@@ -24,9 +24,19 @@ def task_is_assigned_to_organization_user_administers?(user)
...
+      return unless ready_for_dispatch?(root_task.appeal)
+
       create!(assigned_to: BvaDispatch.singleton, parent_id: root_task.id, appeal: root_task.appeal)
     end
 
+    TASK_TYPES_BLOCKING_DISPATCH = [:QualityReviewTask].freeze
+
+    def ready_for_dispatch?(appeal)
+      return false if appeal.tasks.open.where(type: TASK_TYPES_BLOCKING_DISPATCH).any?
+
```
Comment (yoomlam) [inline app/models/tasks/bva_dispatch_task.rb:36]: Leaving room for other checks we should do.

File: app/workflows/complete_case_review.rb [hunk 3]
```diff
@@ -11,7 +11,7 @@ def initialize(case_review_class:, params:)
   def call
     @success = case_review.valid?
 
-    create_quality_review_task if success
+    create_next_task if success
```
Comment (yoomlam) [inline app/workflows/complete_case_review.rb:14]: Renaming method. This method creates a QRTask or BvaDispatchTask.

File: app/workflows/complete_case_review.rb [hunk 4]
```diff
@@ -28,24 +28,32 @@ def review_class
...
-              case_review.task.parent.is_a?(BvaDispatchTask)
+  def create_next_task
+    return if irrelevant_scenario_to_create_next_task?
 
     root_task = case_review.task.root_task
     if QualityReviewCaseSelector.select_case_for_quality_review?
       QualityReviewTask.create_from_root_task(root_task)
-    else
+    elsif open_qr_tasks.blank?
       BvaDispatchTask.create_from_root_task(root_task)
```
Comment (yoomlam) [inline app/workflows/complete_case_review.rb:38]: Only create it BvaDispatchTask if there are no open QualityReviewTask.

File: app/workflows/complete_case_review.rb [hunk 5]
```diff
@@ -28,24 +28,32 @@ def review_class
...
+  # Some situations do not want to move onto the next task:
+  # - Where the appeal is a Legacy appeal
+  # - We are in attorney checkout instead of Judge
+  # - The case review is on a flow that already did QR creation
+  def irrelevant_scenario_to_create_next_task?
+    case_review.appeal.is_a?(LegacyAppeal) ||
+      !case_review.is_a?(JudgeCaseReview) ||
+      case_review.task.parent.is_a?(QualityReviewTask) ||
+      case_review.task.parent.is_a?(BvaDispatchTask)
+  end
```
Comment (yoomlam) [inline app/workflows/complete_case_review.rb:51]: Just moving code. Didn't change semantics.

File: spec/models/tasks/bva_dispatch_task_spec.rb [hunk 6]
```diff
@@ -12,14 +12,16 @@
       end
     end
 
+    let(:root_task) { create(:root_task) }
+    subject { BvaDispatchTask.create_from_root_task(root_task) }
```
Comment (yoomlam) [inline spec/models/tasks/bva_dispatch_task_spec.rb:16]: Refactoring common code to DRY it up.

**Review by jcq — APPROVED** [review]
Ran/examined the specs, everything seems good

File: app/workflows/complete_case_review.rb [hunk 7]
```diff
@@ -28,24 +28,32 @@ def review_class
...
+
+  def open_qr_tasks
+    case_review.task.appeal.tasks.open.where(type: :QualityReviewTask)
+  end
```
[the remark this answers is not in the document]
Comment (yoomlam) [reply app/workflows/complete_case_review.rb]: Updated.

**Pull request merged**